### PR TITLE
corp: add decimal support

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @qiuyesuifeng  PTAL

After this pr, #535 should be resolved.

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @morefreeze

File: src/util/codec/datum.rs [hunk 1]
```diff
@@ -339,21 +373,58 @@ pub fn encode_value(values: &[Datum]) -> Result<Vec<u8>> {
...
+                      Datum::Dec("0.1234".parse().unwrap()),
+                      Datum::Dec("0.0".parse().unwrap()),
+                      Datum::Dec("0".parse().unwrap()),
+                      Datum::Dec("-0.0".parse().unwrap()),
+                      Datum::Dec("-0.0000".parse().unwrap()),
+                      Datum::Dec("-1234.00".parse().unwrap()),
+                      Datum::Dec("-1234".parse().unwrap()),
+                      Datum::Dec("-12.34".parse().unwrap()),
+                      Datum::Dec("-12.340".parse().unwrap()),
+                      Datum::Dec("-0.1234".parse().unwrap())]];
```
Comment (morefreeze) [inline src/util/codec/datum.rs]: Add `".123"` (without leading 0) similar tests?

File: src/util/codec/datum.rs [hunk 2]
```diff
@@ -162,6 +188,7 @@ impl Datum {
             Datum::Bytes(ref bs) => Some(!bs.is_empty() && try!(convert::bytes_to_int(bs)) != 0),
             Datum::Null => None,
             Datum::Dur(d) => Some(!d.is_empty()),
+            Datum::Dec(d) => Some(try!(d.to_f64()) != 0f64),
```
Comment (coocood) [inline src/util/codec/datum.rs]: Need round.

File: src/util/codec/mysql/decimal.rs [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+    }
+
+    /// Get the max needed capacity to encode this decimal.
+    ///
+    /// see also `encode_decimal`.
+    pub fn max_encode_bytes(&self) -> usize {
+        let bits_cnt = self.coeff.bits();
+
+        // 2 * 2 * 2 ≈ 10
+        let mut number_cnt = bits_cnt / 3;
```
Comment (coocood) [inline src/util/codec/mysql/decimal.rs]: `(bit_cnt + 1) / 3` is enough.

File: src/util/codec/mysql/decimal.rs [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+impl FromStr for Decimal {
+    type Err = Error;
+
+    /// Parse the decimal value from string.
+    fn from_str(s: &str) -> Result<Decimal> {
+        let bytes = s.as_bytes();
+        let (mut dot_pos, mut e_pos) = (-1, bytes.len());
+        let mut filtered_bytes = Vec::with_capacity(bytes.len());
+        for (i, &c) in bytes.iter().enumerate() {
+            if (c >= b'0' && c <= b'9') || c == b'-' || c == b'+' {
```
Comment (coocood) [inline src/util/codec/mysql/decimal.rs]: If `-` or `+` doesn't appear at the first byte, it should be invalid decimal.

File: src/util/codec/mysql/decimal.rs (hunk 3 above)
Comment (BusyJay) [inline src/util/codec/mysql/decimal.rs]: No, actually it should be `(bit_cnt + 2) / 3`. I have updated it in the latest version.

File: src/util/codec/mysql/decimal.rs (hunk 4 above)
Comment (BusyJay) [inline src/util/codec/mysql/decimal.rs]: That's not true. It it's positive, `+` can be omitted. See also https://github.com/pingcap/tidb/blob/master/mysql/decimal_test.go#L128.

File: src/util/codec/mysql/decimal.rs [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+        }
+        if self.exp == exp {
+            return self.clone();
+        }
+
+        let to_add = exp - self.exp;
+        let coeff = if to_add > 0 {
+            let divider = BigInt::from(num::pow(BigUint::from(10u8), to_add as usize));
+            let (mut coeff, rem) = self.coeff.div_rem(&divider);
+            if rem >= divider / BigInt::from(2) {
```
Comment (morefreeze) [inline src/util/codec/mysql/decimal.rs]: Is this rounding? Please add some comment.
`rem * BigInt::from(2) >= divider` seems more clear and efficient.

File: src/util/codec/mysql/decimal.rs (hunk 4 above)
Comment (coocood) [inline src/util/codec/mysql/decimal.rs]: I mean '+' or '-' appears in the middle of the decimal string, which is invalid.

Comment (coocood) [inline src/util/codec/mysql/decimal.rs]: Alright.
At last, BigInt from_str will detect it.

File: src/util/codec/mysql/decimal.rs [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+    }
+
+    /// Get the max needed capacity to encode this decimal.
+    ///
+    /// see also `encode_decimal`.
+    pub fn max_encode_bytes(&self) -> usize {
+        let bits_cnt = self.coeff.bits();
+
+        // 2 * 2 * 2 ≈ 10
+        let number_cnt = (bits_cnt + 2) / 3;
```
Comment (morefreeze) [inline src/util/codec/mysql/decimal.rs]: I think 3 is from 2^**3**, but where is (bits_cnt+***2***) from?

File: src/util/codec/mysql/decimal.rs (hunk 5 above)
Comment (BusyJay) [inline src/util/codec/mysql/decimal.rs]: Yes. But I think my version is more straightforward, just as its name indicates "round half up".

File: src/util/codec/mysql/decimal.rs (hunk 6 above)
Comment (BusyJay) [inline src/util/codec/mysql/decimal.rs]: Actually it's a little trick to implement `bits_cnt / 3 + if bits_cnt % 3 > 0 {1} else {0}`.

Comment (morefreeze) [inline src/util/codec/mysql/decimal.rs]: Aha, gotcha. `ceil(bits_cnt/3)`

Comment (coocood) [inline src/util/codec/mysql/decimal.rs]: But the bits_cnt will be at least 8, isn't it?

File: src/util/codec/mysql/decimal.rs [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+        let exp = d.exp + digit_cnt as i32;
+
+        let mut pos = truncated.len();
+        while pos > 0 && truncated[pos - 1] == b'0' {
+            pos -= 1;
+        }
+        truncated = &truncated[..pos];
+
+        if d.is_negative() {
+            try!(self.encode_i64(-exp as i64));
```
Comment (morefreeze) [inline src/util/codec/mysql/decimal.rs]: Sorry, I can't understand this. Please give me more examples.

Comment (morefreeze) [inline src/util/codec/mysql/decimal.rs]: Ignore this.

File: src/util/codec/mysql/decimal.rs (hunk 6 above)
Comment (BusyJay) [inline src/util/codec/mysql/decimal.rs]: No, for example, when coeff is 1, bits_cnt is 1; when coeff is 10, bits_cnt is 4.

**Comment by coocood** [comment]
LGTM

**Comment by morefreeze** [comment]
What's the upper limit of Decimal? Did you test for parsing string like "99999..." ('9' of 30 digits or more)?

**Comment by BusyJay** [comment]
Upper limit has not been added yet. There is a [TODO](https://github.com/pingcap/tikv/pull/582/files#diff-33c81e227baf639277c9d52a8a171ba7R420) for it.

**Comment by morefreeze** [comment]
LGTM

**Comment by BusyJay** [comment]
@ngaut @qiuyesuifeng PTAL

**Pull request merged**